port to mv3

eNV25 opened the pull request:
It works on chromium ~but I haven't tested on firefox. Don't merge.~

Tested on firefox.

eNV25 committed:
port to mv3

Co-authored-by: Eduardo D Sanchez <eduardo.daniel.s.m@gmail.com>

eNV25 commented:
https://developer.chrome.com/docs/extensions/mv3/mv3-migration-checklist/

woodruffw commented:
Awesome, thanks for exploring this. We'll leave it open as a reference for the eventual migration.

eNV25 commented on ff2mpv.js:
```diff
@@ -1,40 +1,33 @@
-}
-
-getOS().then((os) => {
-  var title = os == "win" ? "Play in MP&V" : "Play in MPV (&W)";
```
Was there any point to this code? I didn't really get it, so I removed it.

woodruffw replied:
I have no recollection of why this was added -- my first guess is that someone on Windows wanted a special case for the accelerator key here, since `alt-V` is maybe already occupied by something on Windows.

eNV25 replied:
The current key is `Ctrl+Shift+V`, which I don't really like because that's the same shortcut for pasting in many applications.

Something with `Alt`, like `(Ctrl+)Alt+P` or `(Ctrl+)Alt+M` could work. Or maybe we can just leave it blank. If the use really needs it, they can set it themselves.

woodruffw replied:
Agreed. IMO fine to remove entirely.

eNV25 commented on ff2mpv.js:
```diff
@@ -1,40 +1,33 @@
+  webext.scripting.executeScript({
+    target: { tabId: tabId, allFrames: true },
+    func: () => {
+      document.querySelectorAll("video").forEach((video) => video.pause());
```
The previous code would pause only the first video. This code should pause all.

woodruffw commented:
It looks like MV2's deprecation is happening semi-rapidly after all:

<img width="618" alt="Screenshot 2023-12-05 at 6 24 35 PM" src="https://github.com/woodruffw/ff2mpv/assets/3059210/a2e9f8a6-ff71-43c2-91db-2c1a02062a52">

If I'm reading that right, we'll need to switch to MV3 by June 2024. So we'll probably need to take another look at revitalizing this PR soon 🙂

eNV25 commented:
The only thing that hasn't been resolved so far upstream is background script vs service worker. Chrome only supports service workers and Firefox only supports background scripts.

If it goes on like this we will need to have different `manifest.json` for Chrome and Firefox. The only difference is the `"background"` key.

The extension works equally well with both, so it's not a big problem.

eNV25 commented:
Actually I take that back, apparently you can have both service worker and background script in the same file, according to MDN.

eNV25 commented:
We should probably be using `chrome.runtime.sendMessage()`. @DanSM-5

eNV25 commented:
Here's a possible protocol for `sendMessage`:

`"event"` mandatory key with the following possible values: `"createProfile", "deleteProfile", "updateProfile"`.

`"data"` optional event specific data, for `"createProfile"` and `"updateProfile"` this is the new profile data.

eduardo-sm commented:
> We should probably be using `chrome.runtime.sendMessage()`. @DanSM-5

Oh right, I didn't think about manifest v3 when using `getBackgroundPage`. I used it because it is easier than `chrome.runtime.sendMessage()` but this can be changed. I can  rework it with your proposed protocol.

Overall I need to take a look at the other mv3 specifics:
- chrome.contextMenus.onClicked.addListener()
- chrome.runtime.sendMessage()

Am I missing anything?

> Also this is not exactly my business but I think https://github.com/woodruffw/ff2mpv/pull/108 is over-complicated. I would have just re-used mpv --profile=id, instead of having our own concept of profiles in browser storage.

I tried to use profiles but you still need to send the profile id and while experimenting with it I found more convenient to store the mpv flags. If just limiting to profileId, the logic would be very similar but limited to that specific flag. Using browser sync has the additional benefit that if you use the same account, your configurations are sync and will work (I have to deal with multiple computers, so this helps me). The other part is that you still need to keep opening `mpv.conf` every time you need a change which is what I originally wanted to avoid.

eNV25 commented:
Once you use the correct APIs that I mentioned in `options.js` and `ff2mpv.js`, I can rebase my work here on top of that. I want to avoid touching `options` code too much.

eNV25 commented:
@DanSM-5 Please take a look at https://github.com/eNV25/ff2mpv/blob/pr/mv3/ff2mpv.js. In mv3 all logic should run in event listeners, and all event listeners should be registered in the top level.

In particular I want you to see you the initial `contextMenus` are created at `runtime.onInstalled`. Do you think this will work with your options logic?

DanSM-5 commented:
Sorry for the delay on this.

I have created a PR to your branch @eNV25: https://github.com/eNV25/ff2mpv/pull/1

It is working fine on my side.

eNV25 commented:
Sorry I can't work on it right now. BTRFS decided to corrupt my root filesystem.

DanSM-5 commented:
No hurries on this. Do let me know if this presents issues that you'll like me to take a look.

eNV25 commented:
@DanSM-5 I incorporated your changes, but I wasn't able to properly merge your PR. Moved things around to make the diff look better.

I works according to my testing, please check on your end as well.

DanSM-5 commented:
@eNV25 I see, yesterday I only finished doing the merging with upstream but didn't had a chance to test. I was planning to do that today but I got a bit busy. I'll test this tomorrow.

DanSM-5 reviewed (commented):
Few things I've noticed:

- In options/options.html we need to remove the script tag that adds `browser-polyfill.js`.
- Not sure if this happens to you but in firefox when loading the extension in **about:debugging** it looks like `chrome.runtime.onInstalled.addListener` is not firing for me. I've added logs but those are not recorded and as a result the context menu doesn't work. It works fine if the content is moved to a function that is called when the script runs (similar to how mv2 works)
```javascript
async function load () {
  const profiles = await getProfiles();
 
  if (profiles.length === 0) {
    changeToSingleEntry();
  } else {
    changeToMultiEntries();

    profiles.forEach((profile) => {
      chrome.contextMenus.create({
        parentId: "ff2mpv",
        id: profile.id,
        title: profile.name,
        contexts,
      });
    });
  }
}

load();
```

eNV25 commented:
>     * Not sure if this happens to you but in firefox when loading the extension in **about:debugging** it looks like `chrome.runtime.onInstalled.addListener` is not firing for me. I've added logs but those are not recorded and as a result the context menu doesn't work. It works fine if the content is moved to a function that is called when the script runs (similar to how mv2 works)
> 

Works on my machine. Firefox Developer Edition. All context menus work and update correctly.

The local storage is empty for temporary add-ons, maybe that's what you are seeing? Otherwise have you tried to delete the normal add-on before installing the temporary one?

DanSM-5 commented:
> The local storage is empty for temporary add-ons, maybe that's what you are seeing? Otherwise have you tried to delete the normal add-on before installing the temporary one?

Seems like the issue was to remove the normal ff2mpv. Now it is loading fine.

eNV25 commented on ff2mpv.js:
```diff
@@ -99,75 +100,96 @@ async function createProfile(profile) {
+
+chrome.runtime.onMessage.addListener((request, _sender, sendResponse) => {
+  if (!request) {
+    console.wanr("No request in message");
```
typo

eNV25 commented on ff2mpv.js:
```diff
@@ -1,94 +1,95 @@
-  return browser.runtime.getPlatformInfo().then((i) => i.os);
+function ff2mpv(url, tabId, options = []) {
+  if (tabId != null) {
+    chrome.scripting.executeScript({
```
@woodruffw Should we `await` all of these kinds of calls. Currently we just ignore the `Promise`.

DanSM-5 replied:
@eNV25 I think there is no need for that. The block is to try to pause a video in case there is one playing when you call ff2mpv. The promise rejecting or failing doesn't affect and awaiting it will only slow calling the `sendNativeMessage`.

guest271314 commented on ff2mpv.js:
```diff
@@ -1,94 +1,96 @@
   // Remove single entry
   try {
-    browser.contextMenus.remove('ff2mpv');
+    chrome.contextMenus.remove("ff2mpv");
```
`removeAll()` works.

guest271314 commented on ff2mpv.js:
```diff
@@ -1,94 +1,96 @@
   // Remove sub context menu
   try {
-    browser.contextMenus.remove('ff2mpv');
+    chrome.contextMenus.remove("ff2mpv");
```
`removeAll()` appears to work. As is throws menu item not found.

eNV25 commented:
For a while I've been seeing an issue where the "Play in MPV" submenu, doesn't appear when you first create a profile. Re-creating the profile seems to fix the issue.

DanSM-5 commented:
@eNV25 did it happened to you after the last changes I made?

I noticed the same when testing the suppression of the error `runtime.lastError` and I figure it is because the method remove and removeAll are asynchronous. It should work now that that is being awaited. I didn't get the error again after.

DanSM-5 commented on ff2mpv.js:
```diff
@@ -1,94 +1,92 @@
-  try {
-    browser.contextMenus.remove('ff2mpv');
-  } catch (error) {}
+  chrome.contextMenus.removeAll();
```
You need to await here. `remove` and `removeAll` returns a promise.

DanSM-5 commented on ff2mpv.js:
```diff
@@ -1,94 +1,92 @@
-  try {
-    browser.contextMenus.remove('ff2mpv');
-  } catch (error) {}
+  chrome.contextMenus.removeAll();
```
Similar case here

DanSM-5 commented on ff2mpv.js:
```diff
@@ -99,75 +97,96 @@ async function createProfile(profile) {
+    default:
+      console.warn("No handler for type:", type);
+      return;
+  }
```
We need to call `sendResponse` or it will log an error because the message didn't get a reply.

It could be just this for now as the options is not using the reply.
```javascript
  sendResponse("ok");
```

eNV25 commented:
@DanSM-5 I added await to all `contextMenu` related function calls, and it seems to work now.

eNV25 commented:
I think this is ready. I heard Chrome started disabling mv2 extensions.

eNV25 committed:
Return failure message in message handlers

Co-authored-by: Eduardo D Sanchez <eduardo.daniel.s.m@gmail.com>

DanSM-5 commented:
Yes, Chrome started disabling mv2 in the canary builds about a week ago. Not sure how long it will take to be released on regular chrome but it would be nice to have this merged before that.

DanSM-5 commented:
For the task that is failing it is because of recent changes on eslint v9 that removed that option. Using `eslint@8` should allow to use `--no-eslintrc`.

woodruffw commented on manifest.json:
```diff
@@ -1,47 +1,43 @@
+    "browser_specific_settings": {
         "gecko": {
             "id": "ff2mpv@yossarian.net",
             "strict_min_version": "50.0"
```
I imagine this will now be higher, right? Or could possibly be removed entirely, since only newer FF will support MV3 anyways.

woodruffw reviewed (commented):
LGTM overall, thanks for your hard work on this. I'll do a more in-depth review tonight or tomorrow, time permitting.

eNV25 committed:
manifest: Remove min version for firefox

woodruffw reviewed (approved):
LGTM, thank you for your hard work here! I'll test drive this locally tonight or tomorrow and merge once I confirm it works for me. From there, I'll plan a major release for both Chrome and FF.

woodruffw committed:
Merge branch 'master' into pr/mv3

woodruffw committed:
try and fix the linting

woodruffw committed:
ci: remove js linting entirely

I never checked it anyways.

woodruffw reviewed (approved):
Confirmed that this works locally for me on stable Firefox. Thanks again for your hard work here @eNV25.

woodruffw commented:
I've cut 6.0.0 and released to both AMO and the Chrome Store. It should be live whenever the two review it.

eNV25 commented:
I barely did anything. The original extension was very simple. Most of the work was done by @DanSM-5 who updated his options page changes to mv3.

woodruffw commented:
> I barely did anything. The original extension was very simple. Most of the work was done by @DanSM-5 who updated his options page changes to mv3.

In that case, thanks @DanSM-5 as well!

agusdallalba commented:
The keyboard hotkey is missing, I think it's the `TITLE` that previously was `TITLE = os === "win" ? "Play in MP&V" : "Play in MPV (&W)";` and now is `"Play in MPV"`

eNV25 commented:
![image](https://github.com/user-attachments/assets/65b9c3b7-37e2-4baa-a8d3-20dfcefad7a8)

Dimithrandir commented:
I believe @agusdallalba is referring to the hotkey for the context menu item, not the shortcut for the action button, which now can be customized, as @eNV25 has suggested. 

When the context menu hotkey was removed with release [v6.0.0](https://github.com/woodruffw/ff2mpv/releases/tag/v6.0.0) in #70, I thought it must be because of compatibility issues with manifest v3. But I have tested it now and it works without any issue.

I don't know if it has been removed on purpose, but if it's not, I can create a PR and reintroduce the hotkey for the main context menu. As for individual profiles (context submenu items), users can add hotkeys for them when stating their names, e.g. "Profile &1", "Profile &2", "&4K" etc.

eNV25 commented:
I honestly have no idea what you're talking about. Do have the MDN or Chrome documentation for that feature?

Dimithrandir commented:
I'm sorry, I should have clarified there are some differences between Firefox and Chromium on this topic.

Upon opening a context menu, Firefox will underline a letter in each item which activates the said item. We referred these as **hotkeys**. This is a different feature from the extension shortcut shown in the screenshot by @eNV25.

Chromium will only show a combination of keys (e.g. Ctrl+S, Ctrl+C etc.) for the items that have them defined. However, even if not explicitly shown, Firefox-style hotkeys [are still present and they work](https://askubuntu.com/questions/1303072/how-to-underline-hotkeys-on-google-chrome-or-chromium-right-click-menu).

Also, both Firefox and Chromium seem to automatically assign hotkeys to context menu items added by extensions, usually the initial letter, if it doesn't conflict with another hotkey.

Adding a specific hotkey (called **access key**) for an extension context menu item is documented [in the MDN docs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus/create#title).

This feature [is not documented in the Chrome docs](https://developer.chrome.com/docs/extensions/reference/api/contextMenus#type-CreateProperties), even though it seems to be supported. 

Pre v6.0.0 the context menu item created by ff2mpv in Firefox had a hotkey (**V** for Windows, **W** for Linux). This has been changed with PR #70, as noted by @agusdallalba. I couldn't find the reason, if any, for this decision.

Bellow is how the context menu looks like with the latest release of ff2mpv (no hotkey) and my local fork (with hotkey, same as pre v6.0.0) on Firefox:

<img width="739" height="207" alt="image" src="https://github.com/user-attachments/assets/fd2d7ebb-337e-41df-8cac-1da1d30c37c4" />

agusdallalba commented:
@Dimithrandir thanks a lot for fleshing out the explanation, this indeed is exactly what I meant.

I'm sorry to @eNV25 because instead of explaining I just patched the TITLE variable to say `"Play in MP&V"` and carried on. This was lazy on my part.

In my use case for ff2mpv I never open the youtube (etc.) page, but right click the thumbnail or link I'm interested in and then press V with the same hand to play on mpv, bypassing the website and all its autoplay antipatterns entirely.

Thank you all :)

eNV25 commented:
I discussed this here with @woodruffw: https://github.com/woodruffw/ff2mpv/pull/70#discussion_r1301426384

The maintainer didn't have any recollection of the purpose of the code, so I removed it with his permission.

woodruffw commented:
Ah yeah, this was based on my also not understanding what the hotkeys were vs. shortcuts -- I wasn't using them, so I never hit the behavior change here.

If the lack of context menu hotkeys is an annoyance for some users, I'd be happy to review a PR that re-adds them.

Dimithrandir commented:
It looks rather like it's been just a misunderstanding. I have created #134 , it can be further discussed there.
